Updating rule-design.md (#946)
act-rules/act-rules.github.io: pages/design/rule-design.md

title: Rule Design
---

The WCAG-ACT-RULES-CG rule design builds on WCAG 2.x and its supporting documents. To achieve the goals the following approach is based on [ACT Rules Format 1.0](https://www.w3.org/TR/act-rules-format/):

1. **[Rule properties](#rule-properties)**: Define the test subject and its environment. Includes identifier, name, type and description, as well as other meta data.

2. **[Applicability](#applicability)**: Identify which elements on a page (if any) are to be tested using the rule.

3. **[Expectations](#expectations)**: Assert what must be true about the target elements, in order for them to pass the rule.

4. **[Assumptions](#assumptions)**: Explicitly state all assumptions made by the rule to ensure accountability of the results.

5. **[Accessibility Support](#accessibility-support)**: Provide information on any known feature support issues.

6. **[Background](#background)**: Provide information on relevant resources referenced when developing the rule.

> 7. **[Test cases](#test-cases)**: Define a range of code examples that demonstrate pass, fail and inapplicable outcomes for readers and for validating implementations.

Author: ```suggestion
7. **[Test cases](#test-cases)**: Define a range of code examples that demonstrate pass, fail and inapplicable outcomes for readers and for validating implementations.

8. **Glossary**: A list of any defined terms used within the rule file.
```

## Rule Properties

> A rule file should have a unique hyphenated filename, created from condensing the name and the unique identifier of the rule, _e.g. "aria-attr-defined-5f99a7"_. A rule file should also have a number of metadata properties at the top of the file.

Author: ```suggestion
A rule file will have a unique hyphenated filename, created from condensing the name and the unique identifier of the rule. Ideally this will begin with the test target rather than the expectation, _e.g. "aria-attr-defined-5f99a7"_. 

A rule file should also have a number of metadata properties at the top of the file.
```

> A rule file should have a unique hyphenated filename, created from condensing the name and the unique identifier of the rule, _e.g. "aria-attr-defined-5f99a7"_. A rule file should also have a number of metadata properties at the top of the file.

Author: ```suggestion
A rule file should have a unique hyphenated filename, created from condensing the name and the unique identifier of the rule, usually starting with the test target, _e.g. "aria-attr-defined-5f99a7"_. A rule file should also have the following metadata properties at the top of the file.
```

For more details, see [ACT Rules Format: Structure](https://www.w3.org/TR/act-rules-format/#act-rule-structure).

### Rule identifier

> The 6-digit identifier is created using a random hex generator, such as https://www.browserling.com/tools/random-hex. The identifier remains the same if the rule name changes.

Author: ```suggestion
A 6-digit identifier is created using a random hex generator, such as https://www.browserling.com/tools/random-hex. The identifier remains the same if the rule name changes.
```

### Rule name

> The agreed conventions for naming of rules are as follows:

Author: ```suggestion
The rule name is its descriptive title. The agreed conventions for naming of rules are as follows:
```

- Use sentence case, _e.g. "Heading is descriptive"_, unless using code, _e.g. "`id` attribute value is unique"_, or initials, _e.g. "HTML page has title"_. Note: names starting with a backtick also need to be wrapped with quotes for the parser.
- Use singular tense so that necessary plurals stand out, _e.g. "`video` element has captions"_.
- Front load the applicable thing, _e.g. "Button has accessible name"_.
- Refer to each applicable thing in a consistent way across rules, especially when related or checking the same applicable thing.
- Describe the passing condition, _e.g. "`id` attribute value is unique", rather than "page has no duplicate `id`s"_.
- Do not use hyphens or dashes unless correct for code, _e.g. `aria-*`_.
- Do not use camelCase unless correct for code, _e.g. `iframe` not iFrame, `feImage` in SVG_.
- If unsure, refer to WCAG language used.
- Aim to be unique, which should happen if title includes the applicable thing and passing condition.

### Rule type

The rule type is either atomic or composite.

For more details, see [ACT Rules Format: Rule Types](https://www.w3.org/TR/act-rules-format/#rule-types).

### Description

> A brief description of what the rule does. This should be declarative rather than negative. It should begin with "This rule checks that..." or similar. It should use plain language, and be a well-formed grammatically correct sentence that finishes with a full stop/period.

Author: ```suggestion
The description is a brief synopsis of what the rule does. This should be declarative rather than negative. It should begin with "This rule checks that..." or similar. It should use plain language, and be a well-formed grammatically correct sentence that finishes with a full stop/period.
```

> _For example: "This rule checks that each ARIA state or property has a valid value."_

### Accessibility requirements

> A list of all accessibility requirements that are not satisfied if one or more outcomes of a rule fails. This could be one or more WCAG success criteria, or a WAI-ARIA rule, for example.

Author: ```suggestion
List all accessibility requirements that are not satisfied if one or more outcomes of a rule fails. This could be one or more WCAG success criteria, or a WAI-ARIA rule, for example. Each requirement should indicate the associated conformance level, and map the implications of each outcome, such as "satisfied", "not satisfied", or "further testing needed".
```

title: Rule Design
---

The WCAG-ACT-RULES-CG rule design builds on WCAG 2.x and its supporting documents. To achieve the goals the following approach is based on [ACT Rules Format 1.0](https://www.w3.org/TR/act-rules-format/):

1. **[Rule properties](#rule-properties)**: Define the test subject and its environment. Includes identifier, name, type and description, as well as other meta data.

2. **[Applicability](#applicability)**: Identify which elements on a page (if any) are to be tested using the rule.

3. **[Expectations](#expectations)**: Assert what must be true about the target elements, in order for them to pass the rule.

4. **[Assumptions](#assumptions)**: Explicitly state all assumptions made by the rule to ensure accountability of the results.

> 5. **[Accessibility Support](#accessibility-support)**: Provide information on any known feature support issues.

Author: Seems reasonable and matches the ACT Rules Format wording.

title: Rule Design
---

> The WCAG-ACT-RULES-CG rule design builds on WCAG 2.x and its supporting documents. To achieve the goals the following approach is based on [ACT Rules Format 1.0](https://www.w3.org/TR/act-rules-format/):

Reviewer: WCAG-ACT-RULES-CG can be hyperlinked?

Author: Should it be? It isn't in the current document.

Author: I don't believe a change is needed here, so resolving this comment

> 1. **[Rule properties](#rule-properties)**: Define the test subject and its environment. Includes identifier, name, type and description, as well as other meta data.

Reviewer: Can we refer/ hyperlink to the perhaps a frontmatter section of the act rules format?

2. **[Applicability](#applicability)**: Identify which elements on a page (if any) are to be tested using the rule.

3. **[Expectations](#expectations)**: Assert what must be true about the target elements, in order for them to pass the rule.

4. **[Assumptions](#assumptions)**: Explicitly state all assumptions made by the rule to ensure accountability of the results.

5. **[Accessibility Support](#accessibility-support)**: Provide information on any known feature support issues.

6. **[Background](#background)**: Provide information on relevant resources referenced when developing the rule.

7. **[Test cases](#test-cases)**: Define a range of code examples that demonstrate pass, fail and inapplicable outcomes for readers and for validating implementations.

## Rule Properties

> The rule file should have a unique hyphenated filename, created from condensing the name and the unique identifier, _e.g. "aria-attr-defined-5f99a7"_.

Reviewer: also the `id` should be part of the filename.

Author: Does the wording not already say that? How would you word it to be clearer?

### Rule identifier

The 6-digit identifier is created using a random hex generator, such as https://www.browserling.com/tools/random-hex. The identifier remains the same if the rule name changes.

### Rule name

The agreed conventions for naming of rules are as follows:

- Be succinct, direct and declarative, and avoid unnecessary words, _e.g. an, the_.
- Use declarative verbs, _e.g. has, have, is_.
- Use back ticks to indicate words from code, _e.g. `label` v. Label_.
- Identify code as an element or attribute as appropriate, _e.g. `title` element, `title` attribute_.

> - Use sentence case, _e.g. "Heading is descriptive"_, unless using code, _e.g. "`id` attribute value is unique"_, or initials, _e.g. "HTML page has title"_. Note: names starting with a backtick also need to be wrapped with quotes for the parser. 

Reviewer 2: I would like to see an example of names starting with backticks and wrapped with quotes. Mainly for consistency with the other two examples you provide above.

Author: The second example does that, but it could be difficult when we are quoting the examples. Would this be helpful:

_e.g. "'`id` attribute value is unique'"_

- Use singular tense so that necessary plurals stand out, _e.g. "`video` element has captions"_.
- Front load the applicable thing, _e.g. "Button has accessible name"_.
- Refer to each applicable thing in a consistent way across rules, especially when related or checking the same applicable thing.
- Describe the passing condition, _e.g. "`id` attribute value is unique", rather than "page has no duplicate `id`s"_.
- Do not use hyphens or dashes unless correct for code, _e.g. `aria-*`_.
- Do not use camelCase unless correct for code, _e.g. `iframe` not iFrame, `feImage` in SVG_.
- If unsure, refer to WCAG language used.
- Aim to be unique, which should happen if title includes the applicable thing and passing condition.

### Rule type

> The rule type is either atomic or composite.

Reviewer: link to what is `atomic` or `composite` rule in act format?

- Front load the applicable thing, _e.g. "Button has accessible name"_.
- Refer to each applicable thing in a consistent way across rules, especially when related or checking the same applicable thing.
- Describe the passing condition, _e.g. "`id` attribute value is unique", rather than "page has no duplicate `id`s"_.
- Do not use hyphens or dashes unless correct for code, _e.g. `aria-*`_.
- Do not use camelCase unless correct for code, _e.g. `iframe` not iFrame, `feImage` in SVG_.
- If unsure, refer to WCAG language used.
- Aim to be unique, which should happen if title includes the applicable thing and passing condition.

### Rule type

The rule type is either atomic or composite.

For more details, see [ACT Rules Format: Rule Types](https://www.w3.org/TR/act-rules-format/#rule-types).

### Description

A brief description of what the rule does. This should be declarative rather than negative. It should begin with "This rule checks that..." or similar. It should use plain language, and be a well-formed grammatically correct sentence that finishes with a full stop/period. 

> _For example: "This rule checks that each ARIA state or property has a valid value."_

> ### Accessibility requirements

Reviewer 3: Can you add a note that when linking to WCAG (or other sources with several versions), we use the version number where the stuff we check was first introduced.
https://github.com/act-rules/act-rules.github.io/issues/974#issuecomment-542641648

Author: Note added

> A list of all accessibility requirements that are not satisfied if one or more outcomes of a rule fails. This could be one or more WCAG success criteria, or a WAI-ARIA rule, for example.

Author: ```suggestion
A list of all accessibility requirements that are not satisfied if one or more outcomes of a rule fails. This could be one or more WCAG success criteria, or a WAI-ARIA rule, for example.

**Note**: When linking to WCAG or other sources that may have several versions, use the version number where the relevant requirement was first introduced.
```

## Rule Properties

The rule file should have a unique hyphenated filename, created from condensing the name and the unique identifier.

### Rule identifier

The 6-digit identifier is created using a random hex generator, such as https://www.browserling.com/tools/random-hex. The identifier remains the same if the rule name changes.

### Rule name

> The agreed conventions for naming of rules are as follows:

Reviewer 3: General for this section:
- actually use back ticks where needed 😄 ("`label` element" instead of "label element");
- use "_e.g._" (`_e.g._`) instead of "eg.";
- give full examples when possible (most of the time) (like you do for https://github.com/act-rules/act-rules.github.io/pull/946/files#diff-9772ff397ce17764837b9ae6f1fd71acR39).

- Be succinct, direct and declarative, and avoid unnecessary words, eg. an, the.
- Use declarative verbs, eg. has, have, is.
- Use back ticks to indicate words from code, eg. label v. Label.
- Identify code as an element or attribute, eg. label element.
- Use sentence case, eg. Heading is descriptive, unless using code, eg. id attribute value is unique, or initials, eg. HTML.
- Use singular tense so that necessary plurals stand out, eg. video element has captions.
- Front load the applicable thing, eg. Button has accessible name.
- Refer to each applicable thing in a consistent way across rules, especially when related or checking the same applicable thing.

> - Describe the passing condition, eg. id value is unique, rather than page has no duplicate ids.

Reviewer 3: double quote examples ("`id` value is unique" rather than "page has no duplicate `id`s").

- Do not use hyphens or dashes unless correct for code, eg. aria-*.
- Do not use camelCase unless correct for code, eg. iframe not iFrame, feImage in SVG.
- If unsure, refer to WCAG language used.
- Aim to be unique, which should happen if title includes the applicable thing and passing condition.

### Rule type

The rule type is either atomic or composite.

> 

Reviewer 3: I think we should add something about rule description.
For example: full sentence (ended by a full period), starts with "This rule checks that …", declarative rather than negative.

Author: I'm not sure why it's not there. I thought I made changes to the list at the top of the page also.

Other aspects may be necessary for testing. These can be added as long as they are sufficiently defined.

> For more details, see [ACT Rules Format: Input](https://www.w3.org/TR/act-rules-format/#input)

Reviewer 3: I think ACT-Rules format should be cited upfront. Keep the individual parts where needed, but start by "the rule comply to ACT-Rule format" or something similar.

Author: Presume you mean move this sentence earlier in the page.

### Authors

Names of the Authors. These must be an exact match of names in `contributors` property in `package.json`.

> ## Applicability

Reviewer 3: Both for Applicability and Expectation: recall that they must be correct English sentences (final full period, not skipping article, …)
For example, I think that "Target element has an accessible name" should be "The targe element has an accessible name." (or "The test target has an accessible name.")

Author: OK

Each expectation exposes a reason why an element may not meet a particular conformance requirement. The expectations can be "linked", in that one has to be met before a second can be tested. For example, a rule testing link names may have as its first expectation "Target element has an accessible name", and as a second expectation "Expectation 1 is true for the target element, and the accessible name describes the function of the target element".

For more details, see [ACT Rules Format: Expectations](https://www.w3.org/TR/act-rules-format/#expectations)

## Assumptions

Many accessibility evaluations (especially automated tools) make assumptions about the structure of the web content and the way in which (web) technologies are used. Such assumptions influence the outcome of a test. If the assumptions are made implicitly, it will be difficult to interpret the test result. Comparability and reproduction of results by other tools are limited. Therefore the WCAG-ACT-RULES-CG test include a list of all assumptions, limitations and exceptions made in the design of the rule.

> > _For example:_ A rule for 1.4.1 Use of Color has to make an assumption which CSS-properties are used to make a link visually evident. Typically something like `background`, `border`, `color`, `font`, or `text-decoration`.

Reviewer 3: I like the `>` style of including examples. Maybe that can be generalised to other parts.

Author: Think that comes from the original document. Happy to make it consistent throughout.

While most assumptions relate to the rule itself, there are some assumptions that apply at other stages of the evaluation:

> - It is assumed that the tested web page is the one that has to conform to WCAG 2.0 and that there is no [conforming alternative version](https://www.w3.org/TR/WCAG21/#dfn-conforming-alternate-version).

Reviewer 3: Can you also change WCAG 2.0 to 2.1?
(especially since the link **is** 2.1…)

- It is assumed that the following technologies are accessibility supported: HTML, CSS, WAI-ARIA, ... (See also WCAG-ACT-RULES-CG's [explanation on Accessibility Support](accessibility-support.html)).

For more details, see [ACT Rules Format: Assumptions](https://www.w3.org/TR/act-rules-format/#assumptions)

## Accessibility Support

Information on any known support issues for a feature relevant to the rule.

For more details, see [ACT Rules Format: Accessibility Support](https://www.w3.org/TR/act-rules-format/#accessibility-support)

## Background

> Optional links to information that is relevant reading, such as a success criterion, understanding or technique document, or other specification. 

Reviewer 3: It seems that we are moving toward not that "optional" background, and more structured than now.
Notably, WCAG SC and techniques should be here if used in the rule, grouped by category (there was a PR about updating the background in the templates, but I can't find it anymore…)
I also feel that any external document cited in the rule should appear in the background (#930), which could be mentioned here. But maybe we need to agree on it first 😄

Author: At the moment the ACT Rules format describes "Background" as an optional section. I agree it is rarely omitted in our rules, but also agree that it should be agreed elsewhere before changing that in this page.

For more details, see [ACT Rules Format: Background](https://www.w3.org/TR/act-rules-format/#background)

> ## Test cases

Reviewer 3: Add something about example descriptions. First that they must be here and explain a bit what happen. More ideas on #834 but that's also something that need more work and agreeing first…

Author: Just checking... this is in reference to the line of text between Passed/Failed/Inapplicable Example _n_ and the actual code snippet?

Reviewer 3: Yes.

Other aspects may be necessary for testing. These can be added as long as they are sufficiently defined.

For more details, see [ACT Rules Format: Input](https://www.w3.org/TR/act-rules-format/#input)

### Authors

> Names of the Authors. These must be an exact match of names in `contributors` property in `package.json`.

Reviewer: worth adding a note, that if the contributor does not exist in the package.json, the PR should also amend the contributors list if a new author is intended to be added.

### Authors

Names of the Authors. These must be an exact match of names in `contributors` property in `package.json`. Note: If an author does not exist in `package.json`, they should be added to that file as part of the relevant PR.

## Applicability

Applicability describes which (elements of) web pages should be tested using the rule. These elements are known as test targets. Applicability must be written in plain language, as well-formed grammatically correct sentences, so that it can be used by QA testers. Applicability must rely on well defined properties of the technologies that are tested. For instance, a rule may be applicable to all `video` elements, but it can not be applicable to all `object` elements used to show video, unless the term "video" is further defined.

Use objective, unambiguous definitions within applicability. Finding objective definitions to use in rules can be difficult, if not outright impossible in some cases. The intent here is to ensure repeatability of the rule. Not everything in WCAG testing is entirely repeatable, but when it comes to rule applicability, this is a hard requirement.

> > _For example:_ A rule testing that page titles are descriptive should only apply to specific `title` elements and this could be stated as _"This rule applies to the first HTML title element that is a descendant of the html element of a web page, and contains children that are text nodes that are not only whitespace."_.

Reviewer 3: ```suggestion
> _For example:_ A rule testing that page titles are descriptive should only apply to specific `title` elements and this could be stated as _"This rule applies to the first HTML `title` element that is a descendant of the html element of a web page, and contains children that are text nodes that are not only whitespace."_.
```

Author: Good spot on the back ticks @Jym77 , thanks

- It is assumed that the following technologies are accessibility supported: HTML, CSS, WAI-ARIA, ... (See also WCAG-ACT-RULES-CG's [explanation on Accessibility Support](accessibility-support.html)).

For more details, see [ACT Rules Format: Assumptions](https://www.w3.org/TR/act-rules-format/#assumptions).

## Accessibility Support

Sometimes a feature relevant to a rule is not fully supported by assistive technology and user agents. Include information on any known support issues for a feature relevant to the rule.

For more details, see [ACT Rules Format: Accessibility Support](https://www.w3.org/TR/act-rules-format/#accessibility-support).

## Background

While optional, this provides background information relevant to the development of the rule. Any document cited in the rule should be included here, along with any applicable success criterion, understanding or technique documents, other specifications, and any other relevant reading.

For more details, see [ACT Rules Format: Background](https://www.w3.org/TR/act-rules-format/#background).

## Test cases

> The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **inapplicable** outcomes. All examples should demonstrate good practice, with allowances for omitting code not directly relevant for the rule, so as to be succinct. Additionally, a failing example should only fail in one demonstrated aspect relevant for the rule.

Author: ```suggestion
The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **inapplicable** outcomes, with reasonable coverage of all logically possible cases. All examples should demonstrate good practice, with allowances for omitting code not directly relevant for the rule, so as to be succinct. Additionally, a failing example should clearly fail in only one demonstrated aspect relevant for that rule.
```

> > _For example, a passing example of a page with a `title` element might omit the `lang` attribute on the `html` element, and the `head` and `body` elements in order to be succinct. A failing example might also omit the `title` element. Meanwhile a passing example of a descriptive `title` element would include the `lang` attribute as relevant.

Author: ```suggestion
> _For example, a passing example of a page with a `title` element might omit the `lang` attribute on the `html` element, and the `head` and `body` elements in order to be succinct. A failing example might also omit the `title` element. Meanwhile a passing example of a descriptive `title` element would include the `lang` attribute because it is relevant.
```

> Each test case should be simply and uniquely named, _e.g. "Passed Example 1"_. Each must also include a brief description that explains why the example has the outcome it claims to have.

Reviewer 3: Actually, there are tests in place that check that the examples are named "Passed/Failed/Inapplicable Example X" with the numbers in increasing sequence starting at 1. So the naming requirements are quite stricts.
```suggestion
Each test case must be named, in the format _Passed/Failed/Inapplicable Example X"_ where X is a number, in increasing sequence for each of the three kinds of tests. Each must also include a brief description that explains why the example has the outcome it claims to have.
```

Author: Thanks, I wasn't aware of the checks. I think it would be good to keep the example, so will commit the suggestion and then tweak it.

Other aspects may be necessary for testing. These can be added as long as they are sufficiently defined.

For more details, see [ACT Rules Format: Input](https://www.w3.org/TR/act-rules-format/#input).

> ### Authors

Reviewer 4: This was updated recently, we're now grouping "authors" under an "acknowledgements". Additionally there is a "previous_authros" field too.

Author: Various amendments made to more closely align with ACT Rules Format.

> ### Authors

Author: ```suggestion
### Acknowledgements
```

> Names of the Authors. These must be an exact match of names in `contributors` property in `package.json`. Note: If an author does not exist in `package.json`, they should be added to that file as part of the relevant PR.

Author: ```suggestion
While optional, this can provide information on authors, previous authors, contributors, reviewers and other supporters who have helped with creation of a rule. When listing authors or other acknowledgements, names must be an exact match of names under the `contributors` property in `package.json`. Note: If an author does not exist in `package.json`, they should be added to that file as part of the relevant PR.
```

For more details, see [ACT Rules Format: Accessibility Support](https://www.w3.org/TR/act-rules-format/#accessibility-support).

## Background

While optional, this provides background information relevant to the development of the rule. Any document cited in the rule should be included here, along with any applicable success criterion, understanding or technique documents, other specifications, and any other relevant reading.

For more details, see [ACT Rules Format: Background](https://www.w3.org/TR/act-rules-format/#background).

## Test cases

The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **inapplicable** outcomes, with reasonable coverage of all logically possible cases. All examples should demonstrate good practice, with allowances for omitting code not directly relevant for the rule, so as to be succinct. Additionally, a failing example should clearly fail in only one demonstrated aspect relevant for that rule.

> > \_For example, a passing example of a page with a `title` element might omit the `lang` attribute on the `html` element, and the `head` and `body` elements in order to be succinct. A failing example might also omit the `title` element. Meanwhile a passing example of a descriptive `title` element would include the `lang` attribute because it is relevant.

Author: ```suggestion
> \_For example, a passing example of a page with a `title` element might omit the `lang` attribute on the `html` element, and the `head` and `body` elements in order to be succinct. A failing example might also omit the `title` element. Meanwhile a passing example of a descriptive `title` element would include the `lang` attribute because it is relevant.\_
```

> > \_For example, a passing example of a page with a `title` element might omit the `lang` attribute on the `html` element, and the `head` and `body` elements in order to be succinct. A failing example might also omit the `title` element. Meanwhile a passing example of a descriptive `title` element would include the `lang` attribute because it is relevant.\_

Author: ```suggestion
> _For example_: a passing example of a page with a `title` element might omit the `lang` attribute on the `html` element, and the `head` and `body` elements in order to be succinct. A failing example might also omit the `title` element. Meanwhile a passing example of a descriptive `title` element would include the `lang` attribute because it is relevant.
```

> Each test case must be named, in the format _Passed/Failed/Inapplicable Example X"_ where X is a number, in increasing sequence for each of the three kinds of tests. Each must also include a brief description that explains why the example has the outcome it claims to have.

Author: ```suggestion
Each test case must be named in the format "Passed/Failed/Inapplicable Example X", where X is a number sequentially increasing for each of the three kinds of outcome, _e.g. "Passed Example 1"_. Each must also include a brief description that explains why the example has the outcome it claims to have.
```

The description should:

- use declarative plain language in the format: _This \[test target\] \[meets / does not meet condition\] \[because optional reasons\]._;
- fully qualify the relevant test target, such as a particular element, or element with id, or first such element, or absent element;
- state that the example _has_ or _does not have_ or _contains_ the necessary condition to be met, and optionally any clarifying reasons;
- use sentences that are well-formed, grammatically correct and finish with a full stop/period; and
- use back ticks around any words from code and identify them as elements or attributes.

> > _For example:_ "This page has a `title` element with content."

Author: ```suggestion
> _For example: "This page has a `title` element with content."_
```

Other aspects may be necessary for testing. These can be added as long as they are sufficiently defined.

For more details, see [ACT Rules Format: Input](https://www.w3.org/TR/act-rules-format/#input).

### Authors

Names of the Authors. These must be an exact match of names in `contributors` property in `package.json`. Note: If an author does not exist in `package.json`, they should be added to that file as part of the relevant PR.

> ## Applicability

Reviewer 3: Currently, out of 53 rules we seem to have 30 applicabilities that start with "The rule applies to", 11 start with "This rule applies to" and 12 in other way (including an ugly "The rule only applies to").

=> I think we should make a choice on how to unify this (and then "The rule applies to" seem to be the majority case, my personal preference would rater be "This rule applies to") and document this choice here.

Author: I also prefer "This rule ..." and "This example ..." for these declarative phrases.

Author: Marking this as resolved, as no-one else has commented.

For more details, see [ACT Rules Format: Assumptions](https://www.w3.org/TR/act-rules-format/#assumptions).

## Accessibility Support

Sometimes a feature relevant to a rule is not fully supported by assistive technology and user agents. Include information on any known support issues for a feature relevant to the rule.

For more details, see [ACT Rules Format: Accessibility Support](https://www.w3.org/TR/act-rules-format/#accessibility-support).

## Background

While optional, this provides background information relevant to the development of the rule. Any document cited in the rule should be included here, along with any applicable success criterion, understanding or technique documents, other specifications, and any other relevant reading. 

For more details, see [ACT Rules Format: Background](https://www.w3.org/TR/act-rules-format/#background).

## Test cases

The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **unapplicable** outcomes. 

> Each test case should be simply and uniquely named, _e.g. "Passed Example 1"_. Each example must also include a brief declarative description that uses plain language, and is a well-formed grammatically correct sentence that finishes with a full stop/period. Words from code should use back ticks and be identified as elements or attributes.

Reviewer 3: I would add:
- description should explain why the example has the outcome it claims to have (which is usually not a problem…)
- Use demonstrative form when possible ("this image" vs "the image"); or fully qualified denomination when needed ("the `div` element with an ID of `foo`", if there are several `div` in the example and "this `div`" is not possible; "the first `title` element"; …); or "there is no element with a role of…" (typical for inapplicable examples) (not sure how to name that construction).

Author: Suspect we also need some guidance on the code, around ensuring Passed examples would pass all rules, and the relevant thing in a Failed example being the only thing that would not pass all rules.

@WilcoFiers was there anything else in the AG feedback?

- It is assumed that the following technologies are accessibility supported: HTML, CSS, WAI-ARIA, ... (See also WCAG-ACT-RULES-CG's [explanation on Accessibility Support](accessibility-support.html)).

For more details, see [ACT Rules Format: Assumptions](https://www.w3.org/TR/act-rules-format/#assumptions).

## Accessibility Support

Sometimes a feature relevant to a rule is not fully supported by assistive technology and user agents. Include information on any known support issues for a feature relevant to the rule.

For more details, see [ACT Rules Format: Accessibility Support](https://www.w3.org/TR/act-rules-format/#accessibility-support).

## Background

While optional, this provides background information relevant to the development of the rule. Any document cited in the rule should be included here, along with any applicable success criterion, understanding or technique documents, other specifications, and any other relevant reading. 

For more details, see [ACT Rules Format: Background](https://www.w3.org/TR/act-rules-format/#background).

## Test cases

> The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **inapplicable** outcomes. A passing or inapplicable example should pass other applicable rules. A failing example should only fail in the one aspect being demonstrated.

Author: ```suggestion
The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **inapplicable** outcomes. All examples should demonstrate good practice, with allowances for omitting code not relevant to the rule to retain brevity. Additionally, a failing example should only fail in one demonstrated aspect relevant to the rule.
> _For example, a passing example of a page with a `title` element might omit the `lang` attribute on the `html` element, and the `head` and `body` elements in order to be succinct.
```

> 

Author: Hoping the last adjustment covers this in an appropriate way.

Each test case should be simply and uniquely named, _e.g. "Passed Example 1"_. Each must also include a brief description that explains why the example has the outcome it claims to have. 

The description should:
- use declarative plain language in the format: _This [test target] [meets / does not meet condition] [because optional reasons]._; 

> - fully qualify the relevant test target, such as a particular element, or element with id, or first such element, or absent element, …; 

Author: ```suggestion
- fully qualify the relevant test target, such as a particular element, or element with id, or first such element, or absent element; 
```

- state that the example _has_ or _does not have_ or _contains_ the necessary condition to be met, and optionally any clarifying reasons; 
- use sentences that are well-formed, grammatically correct and finish with a full stop/period; and

> - use back ticks around any words from code and be identify them as elements or attributes.

Author: ```suggestion
- use back ticks around any words from code and identify them as elements or attributes.
```

Other aspects may be necessary for testing. These can be added as long as they are sufficiently defined.

For more details, see [ACT Rules Format: Input](https://www.w3.org/TR/act-rules-format/#input).

### Authors

Names of the Authors. These must be an exact match of names in `contributors` property in `package.json`. Note: If an author does not exist in `package.json`, they should be added to that file as part of the relevant PR.

## Applicability

Applicability describes which (elements of) web pages should be tested using the rule. These elements are known as test targets. Applicability must be written in plain language, as well-formed grammatically correct sentences, so that it can be used by QA testers. Applicability must rely on well defined properties of the technologies that are tested. For instance, a rule may be applicable to all `video` elements, but it can not be applicable to all `object` elements used to show video, unless the term "video" is further defined.

> Finding objective definitions to use in rules can be difficult, if not outright impossible in some cases. The intent here is to ensure repeatability of the rule. Not everything in WCAG testing is entirely repeatable, but when it comes to rule applicability, this is a hard requirement.

Author: ```suggestion
Use objective, unambiguous definitions within applicability. Finding objective definitions to use in rules can be difficult, if not outright impossible in some cases. The intent here is to ensure repeatability of the rule. Not everything in WCAG testing is entirely repeatable, but when it comes to rule applicability, this is a hard requirement.
```

Reviewer 3: 👍

> > _For example:_ A rule testing button accessible names could apply to elements other than the `button` element and this could be stated as _"The rule applies to elements that are included in the accessibility tree with the semantic role of button, except for `input` elements of `type="image"`."_.

Author: ```suggestion
> _For example:_ A rule testing button accessible names could apply to elements other than the `button` element and this could be stated as _"This rule applies to elements that are included in the accessibility tree with the semantic role of button, except for `input` elements of `type="image"`."_.
```

Author: Wondering if I should use a different example, as this particular pair of rules is currently under discussion and may change.

Reviewer 3: I agree, it's fairly easy to pick just any other applicability 😄

Author: ```suggestion
> _For example:_ A rule testing that page titles are descriptive should only apply to specific `title` elements and this could be stated as _"This rule applies to the first HTML title element that is a descendant of the html element of a web page, and
contains children that are text nodes that are not only whitespace."_.
```

For more details, see [ACT Rules Format: Applicability](https://www.w3.org/TR/act-rules-format/#applicability).

## Expectations

The applicability help the testers (or test tools) identify what has to be checked. Following that, the expectations are statements that must be true for the applicable elements to pass the rule. Expectations must be written in plain language, as well-formed grammatically correct sentences, so that it can be used by QA testers. If any of the expectations is false, than the target element failed the rule.

> Each expectation exposes a reason why an element may not meet a particular conformance requirement. The expectations can be "linked", in that one has to be met before a second can be tested.

Author: ```suggestion
Use unambiguous definitions within expectations, even where subjective human judgement may be required to determine conformance. Each expectation exposes a reason why an element may not meet a particular conformance requirement. The expectations can be "linked", in that one has to be met before a second can be tested.
```

Reviewer 3: 👍

- It is assumed that the following technologies are accessibility supported: HTML, CSS, WAI-ARIA, ... (See also WCAG-ACT-RULES-CG's [explanation on Accessibility Support](accessibility-support.html)).

For more details, see [ACT Rules Format: Assumptions](https://www.w3.org/TR/act-rules-format/#assumptions).

## Accessibility Support

Sometimes a feature relevant to a rule is not fully supported by assistive technology and user agents. Include information on any known support issues for a feature relevant to the rule.

For more details, see [ACT Rules Format: Accessibility Support](https://www.w3.org/TR/act-rules-format/#accessibility-support).

## Background

While optional, this provides background information relevant to the development of the rule. Any document cited in the rule should be included here, along with any applicable success criterion, understanding or technique documents, other specifications, and any other relevant reading. 

For more details, see [ACT Rules Format: Background](https://www.w3.org/TR/act-rules-format/#background).

## Test cases

> The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **unapplicable** outcomes. 

Author: ```suggestion
The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **inapplicable** outcomes. A passing or inapplicable example should pass other applicable rules. A failing example should only fail in the one aspect being demonstrated.
```

> The test cases are snippets of code that help with understanding and can be used for validating implementations of the rule. There must be at least one example for **pass**, **fail** and **inapplicable** outcomes. A passing or inapplicable example should pass other applicable rules. A failing example should only fail in the one aspect being demonstrated.

Reviewer 3: I disagree with the last 2 sentences "A passing or inapplicable example should pass other applicable rules. A failing example should only fail in the one aspect being demonstrated."
* Notably, we have several rules that target full documents. With these requirements, the examples must have a `alng` attribute for the `html` element, and a `title` element (or they fail the corresponding rule). I think that is confusing the reader more than helping.
* This also means that adding a new rule might require to update examples in several previous rules because there is now another applicable rule that shouldn't fail. That's just not maintenable and scalable.
* If we split rules on SC as per #1049 we will have elements that fail two rules (while currently, there is one rule per element and each element only fails one rule, which might be failing 2 SC). Requiring that failed examples only fail one rule essentially solves #1049 by saying we can't do it.

Author: How would you suggest we phrase it @Jym77 ? Passing and inapplicable code examples should not include accessibility fails, except possibly by omission for brevity. And failing code examples need to be clear about why the snippet is failing.

Most code snippets are actual snippets, so that is usually possible. For examples where the full page is being checked, then perhaps we should agree a default wrapping code that includes all that is needed by SC while still being brief.

Reviewer 3: I'm not sure…
And I'm also not sure that we enforce "no accessibility fail" for Passed/Inapplicable examples.

We do enforce "do not fail the SC being tested", but I don't think we care about failing other SCs…

@WilcoFiers @kasperisager what is your opinion on Passed/Inapplicable examples failing unrelated SC?
(e.g. should https://act-rules.github.io/rules/97a4e1#inapplicable-example-2 have a `title` element and a `lang` attribute (or other tricks) in order to not fail the corresponding SC?)

Author: Perhaps something about code included in the snippet being an example of good practice, with allowance for brevity in relation to requirements of other SC.

@Jym77 @WilcoFiers @kasperisager I think this is the last outstanding thing to resolve.

Reviewer 3: Yes, I'd prefer it that way.

> Each test case should be simply and uniquely named, _e.g. "Passed Example 1"_. Each example must also include a brief declarative description that uses plain language, and is a well-formed grammatically correct sentence that finishes with a full stop/period. Words from code should use back ticks and be identified as elements or attributes.

Author: ```suggestion
Each test case should be simply and uniquely named, _e.g. "Passed Example 1"_. Each must also include a brief description that explains why the example has the outcome it claims to have. 

The description should:
- use declarative plain language in the format: _This [test target] [meets / does not meet condition] [because optional reasons]._ 
- fully qualify the relevant test target, such as a particular element, or element with id, or first such element, or absent element.
- state that the example _has_ or _does not have_ or _contains_ the necessary condition to be met, and optionally any clarifying reasons. 
- use sentences that are well-formed, grammatically correct and finish with a full stop/period. 
- use back ticks around any words from code and be identify them as elements or attributes.
```

Author: @WilcoFiers @Jym77 does the above suggested change meet your previous comments and what was agreed in #834 ?

I've also suggested a change to the example, and am wondering if a couple of examples would be more helpful.

Reviewer 3: Seems good.
* On the second item "fully qualify…", I think the enumeration is not closed and should end with "…" rather than full period.
* (pedantic typographic mode) if you end your items with full period, the next one should start with an uppercase letter. These items should probably end with a semi-colon instead (except the last one) (and I usually try to do "bla bla**; and**" to emphasise the link between the items (conjunction vs disjonction)).

Author: @Jym77 I've updated the suggestion. Is that what you meant?

Reviewer 3: I've also updated it. It is OK for you?

Author: @Jym77 amended ... typical English phraseology would only include "and" before the last item of a list.

I think lack of other comments can consider this resolved.

Reviewer 3: I'm OK with that.
Hammerring in the "and" is not really useful here.

Minor nerdy concern that this list is semi-colon separated and not full sentences while the list in the "rule name" section is full sentences and period separated (and both look somewhat similar). But I'm OK with the current situation.

Author: I did consider that and decided it was OK to handle them differently. The "rule name" stuff is more complex to explain and much more specific in its requirements.

> > _For example:_ "This page has two `title` elements."

Author: ```suggestion
> _For example:_ "This page has two `title` elements and the first one has content."
```
